web/src/main/kotlin/com/example/ajaxproject/service/GroupChatServiceImpl.kt

```diff
@@ -30,104 +32,89 @@ class GroupChatServiceImpl (
+                groupChatRoomRepository.save(chat).map { user }
+            }
         }
-
-        chat.chatMembers += user
-        logger.info("User with ID {} added to chat ID {}", userId, chatId)
-
-        val updatedChatRoom = groupChatRoomRepository.save(chat)
-
-        return updatedChatRoom.chatMembers
-
     }
 
     @Notification
-    override fun sendMessageToGroup(groupChatDto: GroupChatDto): GroupChatMessageResponse {
-
+    override fun sendMessageToGroup(groupChatDto: GroupChatDto): Mono<GroupChatMessageResponse> {
         val chatMessage = GroupChatMessage(
             id = ObjectId().toHexString(),
             groupChatRoom = groupChatDto.chatId,
             senderId = groupChatDto.senderId,
             message = groupChatDto.message,
             date = Date()
         )
-
-        logger.info("Message sent to group chat ID: {}" , groupChatDto.chatId)
-
-        groupChatMessageRepository.save(chatMessage)
-
-        return toResponseDto(chatMessage)
-
+        return groupChatMessageRepository.save(chatMessage)
+            .doOnSuccess { logger.info("Message sent to group chat ID: ${groupChatDto.chatId}") }
+            .map { toResponseDto(it) }
     }
 
-    override fun getAllGroupMessages(chatId: String): List<GroupChatMessageResponse> {
-
-        val messages = groupChatMessageRepository.findAllMessagesInChat(chatId)
-
-        return messages.map { toResponseDto(it) }
+    @Cacheable("getAllGroupMessages")
+    override fun getAllGroupMessages(chatId: String): Mono<List<GroupChatMessageResponse>> {
+        return groupChatMessageRepository.findAllMessagesInChat(chatId)
+            .map { toResponseDto(it) }
+            .collectList()
+            .cache()
     }
 
-    override fun leaveGroupChat(chatId: String, userId: String): Boolean {
-
-        val user = userService.getUserById(userId)
-
-        val chat = groupChatRoomRepository.findChatRoom(chatId)
-
-        if (chat.adminId != user.id) {
-
-            chat.chatMembers = chat.chatMembers.filterNot { it.id == user.id }
-
-            groupChatRoomRepository.save(chat)
 
-            logger.info("User with ID {} left chat ID {}" ,userId ,chatId)
-
-            return true
-
-        } else {
-
-            logger.warn("Admin attempted to leave chat ID {}" , chatId)
-
-            throw WrongActionException("Admin cant leave a chat")
+    override fun leaveGroupChat(userId: String , chatId: String): Mono<Boolean> {
+        return Mono.zip(
+            groupChatRoomRepository.findChatRoom(chatId),
+            userService.getById(userId)
+        ).flatMap {
+            val chat = it.t1
+            val user = it.t2
+            if (chat.adminId != user.id) {
+                chat.chatMembers = chat.chatMembers.filterNot { it.id == user.id }
+                groupChatRoomRepository.save(chat)
+                    .doOnSuccess { logger.info("User with ID $userId left chat ID $chatId") }
+                    .map { true }
```

Comment:
maybe move doOnSuccess and map out of this flatMap?

```kotlin
return Mono.zip(
...
).flatMap { 
    ...
}
.doOnSuccess(...)
.thenReturn(true)         
``` 
?

Actually, now returning of the boolean doesn't make any sense

Reply:
Yes , i refactored this code . I think it became better